### PR TITLE
[Api][Checkout][Shipping] Shipping method integrity validation

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 1]
```diff
@@ -342,6 +345,25 @@ public function iShouldBeOnTheCheckoutCompleteStep(): void
...
+
+        /** @var Response $response */
+        $response = $this->client->getResponse();
+
+        Assert::same(
+            $this->responseChecker->getValue($response, 'message'),
+            'Given shipment is not eligible for provided shipping method.'
+        );
+
+        Assert::same($response->getStatusCode(), 500);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:364]: I wonder if `500` status code is an appropriate one

Comment (stloyd) [reply src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:364]: `500` should **never** be used by application to inform about something, its meaning is far of the application life & should inform that something went really wrong on level of server architecture (either programming language or server).

This is application information about problem, so it should be `400` (or `409`).

My 2 cents :)

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 2]
```diff
@@ -107,6 +107,10 @@
 
             <itemOperation name="shop_select_shipping_method">
                 <attribute name="method">PATCH</attribute>
+                <attribute name="validation_groups">
+                    <attribute>sylius</attribute>
+                    <attribute>sylius_checkout_choose_shipping_method</attribute>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:112]:
```suggestion
                    <attribute>sylius_checkout_select_shipping_method</attribute>
```
?
to be consistent with the operation name

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 3]
```diff
@@ -342,6 +345,27 @@ public function iShouldBeOnTheCheckoutCompleteStep(): void
...
+        $response = $this->client->getResponse();
+
+        Assert::true($this->isViolationWithMessageInResponse(
+            $response,
+            sprintf('Product does not fit requirements for %s shipping method. Please reselect your shipping method.',
+                $shippingMethod->getName()
+            )
+        ));
+
+        Assert::same($response->getStatusCode(), 400);
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:366]: I would suggest moving this check before the one from line 359 ( comparing two integers has less cost) 🚀

File: src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ (c) Paweł Jędrzejewski
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+-->
+
+<constraint-mapping xmlns="http://symfony.com/schema/dic/constraint-mapping"
+                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                    xsi:schemaLocation="http://symfony.com/schema/dic/constraint-mapping http://symfony.com/schema/dic/services/constraint-mapping-1.0.xsd">
+
+    <class name="Sylius\Bundle\ApiBundle\Command\Checkout\ChooseShippingMethod">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml:14]:
```suggestion
    <class name="Sylius\Bundle\ApiBundle\Command\Checkout\CompleteOrder">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        <constraint name="Sylius\Bundle\ApiBundle\Validator\Constraints\OrderShippingMethodEligibility">
+            <option name="message">sylius.order.shipping_method_eligibility</option>
+            <option name="groups">
+                <value>sylius_checkout_choose_shipping_method</value>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml:18]:
```suggestion
                <value>sylius_checkout_complete</value>
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml:1]: It should be in `Order.xml` I guess

File: src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+<!--
+ This file is part of the Sylius package.
+ (c) Paweł Jędrzejewski
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+-->
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/validation/ShippingMethod.xml:8]:
```suggestion
<!--

 This file is part of the Sylius package.

 (c) Paweł Jędrzejewski

 For the full copyright and license information, please view the LICENSE
 file that was distributed with this source code.

-->
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibility.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+/** @experimental */
+final class OrderShippingMethodEligibility extends Constraint
+{
+    /** @var string */
+    public $message = 'sylius.order.shipping_method_eligibility';
+
+    public function validatedBy(): string
+    {
+        return 'sylius_api_validator_shipping_method_eligibility';
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibility.php:26]:
```suggestion
        return 'sylius_api_validator_order_shipping_method_eligibility';
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        ShipmentRepositoryInterface $shipmentRepository,
+        ShippingMethodRepositoryInterface $shippingMethodRepository,
+        ShippingMethodEligibilityCheckerInterface $eligibilityChecker
+    ) {
+        $this->shipmentRepository = $shipmentRepository;
+        $this->shippingMethodRepository = $shippingMethodRepository;
+        $this->eligibilityChecker = $eligibilityChecker;
+    }
+
+    public function validate($command, Constraint $constraint)
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php:48]:
```suggestion
    public function validate($value, Constraint $constraint)
```
to be consistent with implemented interface

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+use Sylius\Component\Core\Model\ShippingMethodInterface;
+use Sylius\Component\Core\Repository\ShipmentRepositoryInterface;
+use Sylius\Component\Core\Repository\ShippingMethodRepositoryInterface;
+use Sylius\Component\Shipping\Checker\Eligibility\ShippingMethodEligibilityCheckerInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class OrderShippingMethodEligibilityValidator extends ConstraintValidator
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php:27]: You could add a spec for this validator

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        /** @var OrderShippingMethodEligibility $constraint */
+        Assert::isInstanceOf($constraint, OrderShippingMethodEligibility::class);
+
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $value->getOrderTokenValue()]);
+
+        /** @var OrderInterface $order */
+        Assert::isInstanceOf($order, OrderInterface::class);
+
+        /** @var ShipmentInterface $shipment */
+        $shipment = $order->getShipments()->first();
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php:56]: IMO you should check all shipments, not only the first one

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderShippingMethodEligibilityValidatorSpec.php [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    function it_throws_an_exception_if_constraint_does_not_extend_order_token_value_aware_interface(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', ['', new class() extends Constraint {}])
+        ;
+    }
+
+    function it_throws_an_exception_if_constraint_does_not_type_of_order_shipping_method_eligibility(): void {
+        $constraint = new class() extends Constraint implements OrderTokenValueAwareInterface{
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderShippingMethodEligibilityValidatorSpec.php:53]:
```suggestion
        $constraint = new class() extends Constraint implements OrderTokenValueAwareInterface {
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderShippingMethodEligibilityValidatorSpec.php [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+        $orderRepository->findOneBy(['tokenValue' => 'token'])->willReturn(null);
+
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [$value, $constraint])
+        ;
+    }
+
+    function it_add_violation_if_shipment_does_not_match_with_shipping_method(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderShippingMethodEligibilityValidatorSpec.php:88]:
```suggestion
    function it_adds_violation_if_shipment_does_not_match_with_shipping_method(
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderShippingMethodEligibilityValidatorSpec.php [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+                "sylius.order.shipping_method_eligibility",
+                ["%shippingMethodName%" => "InPost"]
+            )
+            ->shouldBeCalled()
+        ;
+
+        $this->validate($value, $constraint);
+    }
+
+    function it_does_not_add_a_violation_if_shipment_match_with_shipping_method(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderShippingMethodEligibilityValidatorSpec.php:124]:
```suggestion
    function it_does_not_add_a_violation_if_shipment_matches_with_shipping_method(
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Tomasz! :1st_place_medal:

**Pull request merged**